src/pool/pool_disjoint.c

```diff
@@ -420,36 +431,26 @@ static void bucket_update_stats(bucket_t *bucket, int in_use, int in_pool) {
         in_pool * bucket_slab_alloc_size(bucket);
 }
 
-static void bucket_decrement_pool(bucket_t *bucket) {
-    bucket_update_stats(bucket, 1, -1);
-    utils_fetch_and_add64(&bucket->shared_limits->total_size,
-                          -(long long)bucket_slab_alloc_size(bucket));
-}
-
 static bool bucket_can_pool(bucket_t *bucket) {
     size_t new_free_slabs_in_bucket;
 
     new_free_slabs_in_bucket = bucket->chunked_slabs_in_pool + 1;
 
     // we keep at most params.capacity slabs in the pool
     if (bucket_max_pooled_slabs(bucket) >= new_free_slabs_in_bucket) {
-        size_t pool_size = 0;
```

Comment:
what was the deadlock exactly? looking at this code, it's not clear

Reply:
I'm not exactly sure how this happened, but the deadlock was caused by two issues: a) decreasing shared_limits->total_size when it was equal to 0, causing it to become SIZE_T_MAX, and b) the utils_compare_exchange() function, whose implementation on Windows does not return a boolean value. Eventually, I decided to remove the while(true) loop entirely.

Reply:
Hm, but isn't decreasing total_size below zero still possible? I don't see any logic here that would cause underflow, so it's probably somewhere else?

Reply:
the code that decreases total_size is here: https://github.com/oneapi-src/unified-memory-framework/pull/1127/files#diff-7d792b84303f5dbbfba19a0b300b3645a0455df415f4246568dc0b1a5b846c73R396 

However, I do not understand how this could happen. This issue was very sporadic and occurred only in the multithreaded Disjoint Pool benchmark on Windows.

Reply:
Hm, okay. I don't think your change should break anything, and if it avoids the deadlock, I'm fine with it. However, perhaps you could add some extra asserts? For example on line 452, where you do utild_fetch_and_sub, you could verify that the original value is big enough (and perhaps there are other places as well).

Reply:
I already added such check here: https://github.com/oneapi-src/unified-memory-framework/pull/1127/files#diff-7d792b84303f5dbbfba19a0b300b3645a0455df415f4246568dc0b1a5b846c73R386

on line 452, I sub the value that I added just a few lines earlier. Since this is atomic, there is no easy way to assert this.

Reply:
Well, you could at least add a check like this:

```
size_t old = utils_fetch_and_sub_u64(&bucket->shared_limits->total_size,
                                    size_to_add);
assert(old >= size_to_add);
```

Reply:
added